### PR TITLE
Make StringValues.ToArray return a copy of internal array

### LINKED ISSUE #36131: StringValues.ToArray does not return a copy if the StringValues contains an array internally.
### Describe the bug

The `StringValues` class is an `IEnumerable<string>`, but it has its own [`ToArray()`](https://github.com/aspnet/Extensions/blob/9eba813df1b11eb171d0b52e8623681e4d555f6e/src/Primitives/src/StringValues.cs#L219) method that shadows the well-known extension method with the same name. This would not be a problem by itself, but **this `ToArray` works surprisingly differently than the extension**: If the StringValues contains an array internally it simply returns that array. This greatly **violates the _principle of least astonishment_** and causes hard-to-localize bugs in user code such as: https://github.com/aspnet/AspNetCore/issues/11016

### To Reproduce
1. Using the latest master version of assembly _Microsoft.Extensions.Primitives_ (the problem existed for a long time though)
2. Run these two tests:
```c#
[Fact]
public void ToArray_ConstructedFromString_CopiesValues()
{
	var singleStringValue = new StringValues("abc");

	var array = singleStringValue.ToArray();
	array[0] = "bcd";

	Assert.Equal("abc", singleStringValue[0]);
}

[Fact]
public void ToArray_ConstructedFromArray_CopiesValues()
{
	var singleStringValue = new StringValues(new [] {"abc"});

	var array = singleStringValue.ToArray();
	array[0] = "bcd";

	Assert.Equal("abc", singleStringValue[0]);
}
```

3. The first one succeeds, as expected.
4. The second one fails, it allows the mutation of `singleStringValue`.

### Expected behavior
Both tests succeed.

### Rationale
1. Seeing a method called `ToArray`, I can't blame anyone that assumes that it creates a copy. I definitely would.
2. This struct is defined as _read only_ at every possible place (`ICollection<string>.IsReadOnly`, all the `NotSupportedException`-s, etc.), it should be immutable as much as practical.
3. Because of the shadowing, the current situation makes this innocent change a breaking chage:
	```
	IEnumerable<string> values = someStringValues;
	var arrayOfValues = values.ToArray();
	Array.Sort(arrayOfValues, StringComparer.Ordinal);
	```
	versus:
	```
	var values = someStringValues;
	var arrayOfValues = values.ToArray();
	Array.Sort(arrayOfValues, StringComparer.Ordinal);
	```
	The first one copies, the second one mutates. Very hard to catch.

### Disclaimer
The proposed fix is a change of behavior in the public-facing api of the `StringValues` class.
However I do not think anyone has used this behavior intentionally because:

- `StringValues` is an `IList` by itself, if someone were to attempt to mutate it they just used that interface (and realized that it does not support mutations intentionally).
- If you really really need a mutable array out of the struct, **it has an implicit `string[]` conversion operator** which does exactly that.

### Further considerations
It should be worth considering whether the `implicit operator string[]` conversion operator and `string[]`-based constructor should create a copy, too. 

I am assuming that these two are implemented without copies for performance reasons, and I left them alone for now in the fix I prepared for the `ToArray` problem. But still, these two points on the API allow simple (and perhaps unintentional) mutation of a struct that is otherwise acts perfectly immutable. It might be worth considering whether the performance gain is worth the integrity risk.

### PR DESCRIPTION
**Pull request by reacheight**
Fixes #36131

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs [hunk 1]
```diff
@@ -286,7 +286,9 @@ private string[] GetArrayValue()
             object value = _values;
             if (value is string[] values)
             {
-                return values;
+                var valuesCopy = new string[values.Length];
```
Comment (GrabYourPitchforks) [inline src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs:289]: FYI the canonical way to write this is `T[] copy = (T[])oldArray.Clone();`.

File: src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs [hunk 2]
```diff
@@ -286,7 +286,8 @@ private string[] GetArrayValue()
             object value = _values;
             if (value is string[] values)
             {
-                return values;
+                var valuesCopy = (string[])values.Clone();
+                return valuesCopy;
```
Comment (jkotas) [inline src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs:290]:
```suggestion
                return (string[])values.Clone();
```

**Comment by davidfowl** [comment]
I'm 85% sure we didn't agree to fix this. We want to understand the performance impact to ASP.NET Core before we make this change.

**Comment by GrabYourPitchforks** [comment]
@davidfowl But there's still that 15% of you that longs for adventure! 🤠

**Comment by kevingosse** [comment]
I feel like this is a naming issue more than an undesirable behavior. If this method is changed, I really wish another allocation-free mechanism is provided to fetch the headers, probably with a method name that more explicitly describes the implications (a bit like `MemoryStream.GetBuffer`.

**Comment by stephentoub** [comment]
@reacheight, thanks for working on this.

@davidfowl, @maryamariyan, sounds like this and #36131 should be closed?

**Comment by maryamariyan** [comment]
Closing as per conversation thread.

**Comment by GrabYourPitchforks** [comment]
FYI, the original bug https://github.com/dotnet/aspnetcore/issues/11016 (which became https://github.com/dotnet/runtime/issues/36131) is still active. We need to resolve that bug somehow now that this PR is rejected.

**Pull request closed without merging**